File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> direction = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

int search(int row, int col, int n, int m, vector<vector<bool>>& visited, vector<string>& maps){
    if(row < 0 || row >= n || col < 0 || col >= m) return 0;
    if(visited[row][col]) return 0;
    if(maps[row][col] == 'X') return 0;
    
    visited[row][col] = true;
    int result = maps[row][col] - '0';
    for(vector<int> d : direction){
        result += search(row + d[0], col + d[1], n, m, visited, maps);
    }
    return result;
}

vector<int> solution(vector<string> maps) {
    int n = maps.size();
    int m = maps[0].size();
    vector<vector<bool>> visited(n, vector<bool>(m, false));
    
    vector<int> answer;
    for(int r = 0; r < n; r++){
        for(int c = 0; c < m; c++){
            int days = search(r, c, n, m, visited, maps);
            if(days > 0) answer.push_back(days);
        }
    }
    
    if(answer.empty()) return {-1};
    sort(answer.begin(), answer.end());
    return answer;
}

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
import sys
sys.setrecursionlimit(10000) # 재귀 한도 증가

direction = [[-1, 0], [0, 1], [1, 0], [0, -1]]

def search(row, col, n, m, visited, maps):
    if row < 0 or row >= n or col < 0 or col >= m:
        return 0
    if maps[row][col] == "X":
        return 0
    if visited[row][col]:
        return 0
    
    visited[row][col] = True
    result = int(maps[row][col])
    for d in direction:
        result += search(row + d[0], col + d[1], n, m, visited, maps)
    
    return result

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    visited = [([False] * m) for i in range(n)]
    
    answer = []
    for r in range(n):
        for c in range(m):
            days = search(r, c, n, m, visited, maps)
            if days > 0:
                answer.append(days)
    
    answer.sort()
    return answer if answer else [-1]